File: 02_Least_Change.py
# In this problem we will develope a solution to finding the least amount
# coins required to give change to a user





def least_change(current_change, coins, cache):
	if(current_change == 0):
		return 0

	min = current_change	

	for coin in coins:
		if (current_change - coin) >= 0:
			if current_change - coin in cache[0]:
				coin_number = cache[0][current_change - coin]
			else:
				coin_number = least_change(current_change - coin, coins, cache)
				cache[0][current_change - coin] = coin_number

			# Find the min of the change below, +1 to include this
			if min > coin_number + 1:
				min = coin_number + 1
				min = coin_number + 1

	return min

def _least_change(current_change, coins):
	if(current_change == 0):
		return 0

	min = current_change	

	for coin in coins:
		if (current_change - coin) >= 0:
			coin_number = _least_change(current_change - coin, coins)

			# Find the min of the change below, +1 to include this
			if min > coin_number + 1:
				min = coin_number + 1
				min = coin_number + 1

	return min

def main():
	coins = [25, 10, 5, 1]
	change = 12345
	least = least_change(change, coins, [{}])

	print "Least change for %d is %d" % (change, least)


if __name__ == '__main__':
	main()

File: 02_Least_Change.java
class LeastChange{

	public int leastChange(int current_change, int[] coins){
		int[] cache = new int[current_change];
		for (int i = 1; i < cache.length; i++) cache[i] = -1;
		return leastChange(current_change, coins, cache); 
	}

	public int leastChange(int current_change, int[] coins, int[] cache){
		if(current_change == 0){
			return 0;
		}

		int min = current_change;
		for(int coin : coins){
			// If the number has been cached
			if(current_change - coin >= 0){
				int coin_count;
				if(cache[current_change - coin] >= 0){
					coin_count = cache[current_change - coin];
				}else{
					coin_count = leastChange(current_change - coin, coins, cache);
					cache[current_change - coin] = coin_count;
				}

				if(min > coin_count + 1) min = coin_count + 1;
			}
		}

		return min;
	}


	public static void main(String[] args){
		int[] coins = {25, 10, 5, 1};
		int change = 123456;

		LeastChange a = new LeastChange();
		int least = a.leastChange(change, coins);
		System.out.printf("Least change for %d is %d\n", change, least);
	}
}